blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md

```diff
@@ -0,0 +1,97 @@
+
+# Test Input Initial Encapsulation
+
+Let's extract `Transaction` creation code into `TestTransactions`
+
+:include-file: com/example/portfolio/ProfitCalculatorWithBasicEncapsulationTest.java {title: "ProfitCalculatorTest.java with initial encapsulation"}
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransaction", entry: "createTransaction"}
+
+Less boilerplate and a good start.
+
+Question: Can you still spot the problem?
+
+--TODO-- add reveal plugin to hide the answer and show it only on a click
+
+As we add more properties to `Transaction` class, some of the tests may need to set extra transaction properties. 
+Additionally some of the properties will be completely irrelevant to the business logic under test, yet you will be forced to set them anyway.
+
+# Test Input Encapsulation With Webtau Table Data   
+
+I was dealing with test problems like above for awhile and eventually came up with a solution that fits my needs.
+I hope it will fit your needs as well. The solution is to use flexible data structure like [Webtau TableData](https://twosigma.github.io/webtau/guide/reference/table-data)
+
+:include-file: com/example/portfolio/ProfitCalculatorWithTableDataTest.java {title: "ProfitCalculatorTest.java with TableData encapsulation", commentsType: "inline"}
+
+Note: [Webtau](https://github.com/twosigma/webtau) open source project started as my answer to common testing problems I encountered. 
+While I am not working at [Two Sigma](https://www.twosigma.com/) anymore I am still contributing and using it on a regular basis.     
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransactions", commentsType: "inline", entries: ["createTransactions", "genTransactionId"]}
+
+Notice how `createTransactions` defaults values when they are not present?
+As a result of this, tests that don't need say `id` or `lot` are free to ignore them.
+
+:include-java: com/example/portfolio/ProfitCalculatorWithTableDataNoIdNoLotTest.java {title: "ignored id,lot Transaction properties", commentsType: "inline", entry: "profitShouldBeZeroIfNoLotsSet"}
+
+If later a new required property will be added to `Transaction`, you won't have to change all your existing tests. 
+Instead you will update `createTransaction` with a new default value. 
+The only time you will have to change the tests to use the new property is if the new property is affecting the logic under test.
+
+# Try it out
+
+:include-file: webtau-core-java-dep.xml {title: "Maven dependency"}
+
+* [Webtau user guide](https://twosigma.github.io/webtau/guide/)
+* [Webtau github](https://github.com/twosigma/webtau)
+
+# Summary
+
+* Tests input preparation can be verbose and brittle.
+* Identify core domain objects and provide a convenient way to create them.
+* Use as little data as necessary to prove logic correctness. Default the rest. 
```

Comment:
Per our call, this idea is the heart of the article, and deserves it's own paragraph